src/core/event_engine/effect_dispatcher.rs

```diff
@@ -0,0 +1,260 @@
+use crate::{
+    core::event_engine::{Effect, EffectHandler, EffectInvocation},
+    lib::alloc::{rc::Rc, vec, vec::Vec},
+};
+use phantom_type::PhantomType;
+use spin::rwlock::RwLock;
+
+/// State machine effects dispatcher.
+#[allow(dead_code)]
+pub(crate) struct EffectDispatcher<EH, EF, EI>
+where
+    EI: EffectInvocation<Effect = EF>,
+    EH: EffectHandler<EI, EF>,
+    EF: Effect<Invocation = EI>,
+{
+    /// Effect invocation handler.
+    ///
+    /// Handler responsible for providing actual implementation of
+    handler: EH,
+
+    /// Dispatched effects managed by dispatcher.
+    ///
+    /// There are effects whose lifetime should be managed by the dispatcher.
+    /// State machines may have some effects that are exclusive and can only run
+    /// one type of them at once. The dispatcher handles such effects
+    /// and cancels them when required.
+    managed: RwLock<Vec<Rc<EF>>>,
+
+    _invocation: PhantomType<EI>,
+}
+
+impl<EH, EF, EI> EffectDispatcher<EH, EF, EI>
+where
+    EI: EffectInvocation<Effect = EF>,
+    EH: EffectHandler<EI, EF>,
+    EF: Effect<Invocation = EI>,
+{
+    /// Create new effects dispatcher.
+    pub fn new(handler: EH) -> Self {
+        EffectDispatcher {
+            handler,
+            managed: RwLock::new(vec![]),
+            _invocation: Default::default(),
+        }
+    }
+
+    /// Dispatch effect associated with `invocation`.
+    pub fn dispatch<F>(&self, invocation: &EI, mut f: F)
+    where
+        F: FnMut(Option<Vec<EI::Event>>),
+    {
+        if let Some(effect) = self.handler.create(invocation) {
+            let effect = Rc::new(effect);
+
+            if invocation.managed() {
+                let mut managed = self.managed.write();
+                managed.push(effect.clone());
+            }
+
+            // Placeholder for effect invocation.
+            effect.run(|events| {
+                // Try remove effect from list of managed.
+                self.remove_managed_effect(&effect);
+
+                // Notify about effect run completion.
+                // Placeholder for effect events processing (pass to effects handler).
+                f(events);
```

Comment:
We don't have implemented effects yet, but I assume that some of them would return events, which should be processed by the event engine.
It is exactly as it says – it is a _placeholder_ for functionality, which I expect to appear with implemented effects.
The function pointer can be passed to the created effect, so it will be called from it.

Reply:
I'm still not sure about that. 
Anyway imo you can merge that. 
Rest of codes looks good.